### PR TITLE
Add factory and update config for tail sampling processor

### PR DESCRIPTION
**Pull request by songy23**
Updates #146

### CODE REVIEW

**Comment by songy23** [comment]
I'm not familiar with the tail sampling policies part so may need help from you @pjanotti @tigrannajaryan.

File: processor/tailsampling/processor.go [hunk 1]
```diff
@@ -69,40 +72,41 @@ const (
...
 	}
 
-	for _, policy := range policies {
-		policyCtx, err := tag.New(tsp.ctx, tag.Upsert(tagPolicyKey, policy.Name), tag.Upsert(observability.TagKeyReceiver, sourceFormat))
-		if err != nil {
-			return nil, err
-		}
-		policy.ctx = policyCtx
-	}
+	// TODO(#146): add policies to TailBasedCfg
```
Comment (pjanotti) [inline processor/tailsampling/processor.go]: Policies have an interface and we will need to add the config factories for each one (as a temporary alternative we can use the old code to read them but I want to double-check that before going in that direction).

File: processor/tailsampling/factory_test.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+	"github.com/stretchr/testify/require"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-service/exporter/exportertest"
+	"github.com/open-telemetry/opentelemetry-service/service/builder"
+)
+
+func TestCreateProcessor(t *testing.T) {
+	factory := &Factory{}
+	require.NotNil(t, factory)
```
[the remark this answers is not in the document]
Comment (songy23) [reply processor/tailsampling/factory_test.go]: Good catch, removed.

File: processor/tailsampling/processor.go (hunk 1 above)
Comment (songy23) [reply processor/tailsampling/processor.go]: Got it, either way I would prefer to add it in a follow-up PR.

**Comment by MohamedElqdusy** [comment]
@songy23 config tests are missing, aren't they?

**Comment by songy23** [comment]
Config is already in https://github.com/open-telemetry/opentelemetry-service/blob/master/service/builder/sampling_builder.go (no tests though). I'll add some tests for them.

**Comment by songy23** [comment]
Added tests in https://github.com/open-telemetry/opentelemetry-service/commit/4b744c0975fe46aad510b73c8fa31fed6c683363, PTAL.

**Comment by songy23** [comment]
Looks like we may need to do some more clean-ups on the old config. Currently tail sampling and probabilistic sampling can be configured in two ways:
1. Through "processors" in config, which I think is the expected approach.
2. Through "sampling". E.g 
https://github.com/open-telemetry/opentelemetry-service/blob/4c7c3ca57110c32803b56cec3fe4733afd74c952/service/builder/testdata/sampling_config.yaml#L27-L66
I think this is for backwards-compatibility. Maybe we should just remove the entire "sampling" section, since tail sampling and probabilistic sampling should just be treated as normal processors? WDYT @pjanotti @tigrannajaryan @MohamedElqdusy

**Comment by pjanotti** [comment]
@songy23 that code under builder should be leftover from the port from OC, currently it shouldn't be used by OTelSvc - we have to remove it.

**Comment by songy23** [comment]
That makes sense - will do in a separate PR :)

File: service/builder/sampling_builder.go [hunk 3]
```diff
@@ -159,20 +161,16 @@ func (sCfg *SamplingCfg) InitFromViper(v *viper.Viper) *SamplingCfg {
 
 // TailBasedCfg holds the configuration for tail-based sampling.
 type TailBasedCfg struct {
```
Comment (MohamedElqdusy) [inline service/builder/sampling_builder.go]: nit: config as a name should be consistent with other processors.

File: processor/tailsampling/factory.go [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+// Type gets the type of the config created by this factory.
+func (f *Factory) Type() string {
+	return typeStr
+}
+
+// CreateDefaultConfig creates the default configuration for processor.
+func (f *Factory) CreateDefaultConfig() configmodels.Processor {
+	return &builder.TailBasedCfg{
+		DecisionWait: 30 * time.Second,
+		NumTraces:    50000,
```
Comment (MohamedElqdusy) [inline processor/tailsampling/factory.go]: Is there a reason for those numbers?

File: service/builder/sampling_builder.go (hunk 3 above)
Comment (songy23) [reply service/builder/sampling_builder.go]: Will rename it when I removed the old config and moved it to tailsampling/config.go (in the follow-up PR).

File: processor/tailsampling/factory.go (hunk 4 above)
Comment (songy23) [reply processor/tailsampling/factory.go]: I'm not sure either. This is copied from https://github.com/open-telemetry/opentelemetry-service/blob/4c7c3ca57110c32803b56cec3fe4733afd74c952/service/builder/sampling_builder.go#L170-L176
@pjanotti any idea on those numbers?

Comment (pjanotti) [reply processor/tailsampling/factory.go]: These are arbitrary and there is no good criteria other than "reasonableness", that I am aware :), to select those defaults.

File: processor/tailsampling/processor.go [hunk 5]
```diff
@@ -69,40 +72,41 @@ const (
...
+
 	tsp := &tailSamplingSpanProcessor{
 		ctx:             context.Background(),
-		maxNumTraces:    maxNumTraces,
-		policies:        policies,
+		nextConsumer:    nextConsumer,
+		maxNumTraces:    cfg.NumTraces,
 		logger:          logger,
 		decisionBatcher: inBatcher,
+		// policies:        policies,
```
Comment (tigrannajaryan) [inline processor/tailsampling/processor.go]: Please remove commented code. I do not know if we this called in contribution guidelines, if not I believe we should.

File: processor/tailsampling/processor.go [hunk 6]
```diff
@@ -69,40 +72,41 @@ const (
...
 
-	for _, policy := range policies {
-		policyCtx, err := tag.New(tsp.ctx, tag.Upsert(tagPolicyKey, policy.Name), tag.Upsert(observability.TagKeyReceiver, sourceFormat))
-		if err != nil {
-			return nil, err
-		}
-		policy.ctx = policyCtx
-	}
+	// TODO(#146): add policies to TailBasedCfg
+	// for _, policy := range policies {
```
Comment (tigrannajaryan) [inline processor/tailsampling/processor.go]: Same here, commented code.

File: processor/tailsampling/processor.go [hunk 7]
```diff
@@ -69,40 +72,41 @@ const (
...
-// a trace should be sampled or not. Providing expectedNewTracesPerSec helps with allocating data structures
-// with closer to actual usage size.
-func NewTailSamplingSpanProcessor(
-	policies []*Policy,
-	maxNumTraces, expectedNewTracesPerSec uint64,
-	decisionWait time.Duration,
-	logger *zap.Logger) (consumer.TraceConsumer, error) {
+// NewTraceProcessor returns a processor.TraceProcessor that will perform tail sampling according to the given
+// configuration.
+func NewTraceProcessor(logger *zap.Logger, nextConsumer consumer.TraceConsumer, cfg builder.TailBasedCfg) (processor.TraceProcessor, error) {
```
Comment (tigrannajaryan) [inline processor/tailsampling/processor.go]: Is there a good reason that `TailBasedCfg` is in `builder` package? If not then perhaps it should be in this package?
Actually, I am not sure `sampling_builder.go` should exist in `builder` package at all, but that's probably a topic for a separate cleanup PR.

File: service/builder/sampling_builder.go (hunk 3 above)
Comment (tigrannajaryan) [reply service/builder/sampling_builder.go]: Sounds good, that's good to know.

File: service/builder/sampling_builder.go [hunk 8]
```diff
@@ -159,20 +161,16 @@ func (sCfg *SamplingCfg) InitFromViper(v *viper.Viper) *SamplingCfg {
...
-
-// NewDefaultTailBasedCfg creates a TailBasedCfg with the default values.
-func NewDefaultTailBasedCfg() *TailBasedCfg {
-	return &TailBasedCfg{
-		DecisionWait: 30 * time.Second,
-		NumTraces:    50000,
-	}
+	// ExpectedNewTracesPerSec sets the expected number of new traces sending to the tail sampling processor
+	// per second. This helps with allocating data structures with closer to actual usage size.
+	ExpectedNewTracesPerSec uint64 `mapstructure:"expected-new-traces-per-sec"`
```
Comment (tigrannajaryan) [inline service/builder/sampling_builder.go]: Is this a new config setting? I cannot find it in the old implementation.

File: processor/tailsampling/processor.go (hunk 7 above)
Comment (songy23) [reply processor/tailsampling/processor.go]: I think that config is inherited from opencensus-service where sampling is configured through builder. I'll move it to processor/tailsampling and clean up the old configs in the next PR.

File: processor/tailsampling/processor.go (hunk 5 above)
Comment (songy23) [reply processor/tailsampling/processor.go]: Removed. BTW it's not included in the guidelines.

File: processor/tailsampling/processor.go (hunk 6 above)
Comment (songy23) [reply processor/tailsampling/processor.go]: Done.

File: service/builder/sampling_builder.go (hunk 8 above)
Comment (songy23) [reply service/builder/sampling_builder.go]: It's not in the old config but it's required by the tail sampler (even before this PR): 
https://github.com/open-telemetry/opentelemetry-service/blob/e5d25031e9a49d14d3dd890f4b0138bdaf5b82ab/internal/collector/processor/tailsampling/processor.go#L74-L82

**Pull request merged**